File: src/raw-pipeline/raw-core/src/types/adjustment/crop.rs
//! `Crop` geometry type — normalised crop rect + straighten angle.
//!
//! Extracted from `mod.rs` to keep that file under the 600-LOC hard budget
//! (#772). Re-exported as `adjustment::Crop` via `pub use crop::Crop`.

/// Geometry (crop + straighten / rotate) per spec § 3.12.
///
/// All four edge fields are normalised to `[0, 1]` relative to the
/// display-oriented (post-EXIF orientation) image dimensions. Identity
/// (no-crop) is the full frame: `top=0, left=0, bottom=1, right=1, angle=0`.
/// The XMP boolean `crs:HasCrop` is derived — emitted only when
/// [`Crop::is_identity`] returns `false`.
///
/// `angle` is in degrees, positive = clockwise (reference-renderer
/// convention). The rotated-image-then-extract-rect math lives in
/// `stages::crop`. Exact 90 / 180 / 270 are taken when paired with the
/// matching orthogonal crop rect.
#[derive(Clone, Copy, Debug, PartialEq)]
pub struct Crop {
    /// Top edge, normalised to `[0, 1]` of display-oriented image height.
    /// `0.0` = top of frame; `1.0` = bottom. Default (identity) = `0.0`.
    pub top: f32,
    /// Left edge, normalised to `[0, 1]` of display-oriented image width.
    /// Default (identity) = `0.0`.
    pub left: f32,
    /// Bottom edge, normalised to `[0, 1]`. Default (identity) = `1.0`.
    pub bottom: f32,
    /// Right edge, normalised to `[0, 1]`. Default (identity) = `1.0`.
    pub right: f32,
    /// Straighten rotation in degrees, positive = clockwise (reference-renderer
    /// convention). Default = `0.0`. Off-axis angles bilinear-resample; exact
    /// 90 / 180 / 270 are taken when paired with the matching orthogonal crop rect.
    pub angle: f32,
}

impl Crop {
    /// Identity crop: full frame, zero rotation. Equivalent to `Default::default()`
    /// but available in `const` contexts.
    pub const IDENTITY: Self = Self {
        top: 0.0,
        left: 0.0,
        bottom: 1.0,
        right: 1.0,
        angle: 0.0,
    };

    /// True iff this crop is the full-frame, zero-rotation identity. Used by
    /// the XMP serializer (omit the `crs:Crop*` group entirely) and by the
    /// `stages::crop` early-exit path so identity crops cost nothing.
    ///
    /// Equality is exact — defaults are concrete `0.0` / `1.0` literals that
    /// round-trip through XMP without drift.
    pub fn is_identity(&self) -> bool {
        self.top == 0.0
            && self.left == 0.0
            && self.bottom == 1.0
            && self.right == 1.0
            && self.angle == 0.0
    }

    /// True iff the rect (ignoring rotation) is well-formed: every edge in
    /// `[0, 1]`, `right > left`, `bottom > top`. Inverted or empty rects
    /// per spec § 3.12 are invalid; `stages::crop` treats them as identity.
    pub fn rect_is_valid(&self) -> bool {
        (0.0..=1.0).contains(&self.top)
            && (0.0..=1.0).contains(&self.left)
            && (0.0..=1.0).contains(&self.bottom)
            && (0.0..=1.0).contains(&self.right)
            && self.right > self.left
            && self.bottom > self.top
    }
}

impl Default for Crop {
    fn default() -> Self {
        Self::IDENTITY
    }
}

File: src/apple/Packages/MapleCore/Sources/MapleCore/Crop.swift
// Crop.swift — geometry (crop + straighten) value type for `AdjustmentModel`.
//
// Split out of `AdjustmentModel.swift` in #366 to keep that file under
// CONTRIBUTING.md's 600-line hard budget once the four point-curve fields
// landed on the model. Pure move — the type is unchanged.

import Foundation

/// Geometry (crop + straighten) per spec § 3.12 / ticket #277. Mirror of
/// `raw_core::types::Crop`. Coordinates are normalised to `[0, 1]` against
/// the display-oriented image dimensions (post-EXIF rotation).
///
/// The XMP wire format is gated by `crs:HasCrop` — the serializer emits the
/// `crs:Crop*` group only when `isIdentity` is false, and the parser
/// silently drops a stale `crs:Crop*` rect when the marker is absent.
public struct Crop: Codable, Sendable, Equatable, Hashable {
    public var top: Double
    public var left: Double
    public var bottom: Double
    public var right: Double
    /// Straighten rotation in degrees, positive = clockwise (reference-renderer convention).
    public var angle: Double

    public init(
        top: Double = 0,
        left: Double = 0,
        bottom: Double = 1,
        right: Double = 1,
        angle: Double = 0
    ) {
        self.top = top
        self.left = left
        self.bottom = bottom
        self.right = right
        self.angle = angle
    }

    /// Identity (full frame, no rotation).
    public static let identity = Crop()

    /// True when the crop is the full-frame, zero-rotation identity. Used
    /// by the XMP serializer to omit the `crs:Crop*` group entirely.
    public var isIdentity: Bool {
        top == 0 && left == 0 && bottom == 1 && right == 1 && angle == 0
    }

    /// True when the rect (ignoring rotation) is well-formed: every edge
    /// in `[0, 1]` with `right > left` and `bottom > top`. Inverted or
    /// empty rects per spec § 3.12 are invalid and the renderer falls back
    /// to identity.
    public var rectIsValid: Bool {
        (0...1).contains(top)
            && (0...1).contains(left)
            && (0...1).contains(bottom)
            && (0...1).contains(right)
            && right > left
            && bottom > top
    }
}
